<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

In [8]:
print("[", end = "")
for number in range(10):
    if (number % 2 == 0):
        continue
    if (number == 9):
        print(f"{number**2}", end = "")
    else:
        print(f"{number**2},", end = " ")
print("]")

[1, 9, 25, 49, 81]


In [10]:
%%timeit
print("[", end = "")
for number in range(10):
    if (number % 2 == 0):
        continue
    if (number == 9):
        print(f"{number**2}", end = "")
    else:
        print(f"{number**2},", end = " ")
print("]")

[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 4

In [3]:
[odd_number**2 for odd_number in range(10) if odd_number % 2 != 0]

[1, 9, 25, 49, 81]

In [14]:
%%timeit
[odd_number**2 for odd_number in range(10) if odd_number % 2 != 0]

626 ns ± 37.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


The method used in exercise 1 has a mean time of 30.4 μs while the list comprehension method in exercise 2 has a mean time of 626 ns. Hence, the list compresion method is faster.

In [18]:
sample_q, time = 100, 0
while sample_q > 50:
    sample_q = sample_q * 95/100  #reducing sample quantity by 5%
    time += 1
print(f"Time taken for sample to reduce to half its original amount = {time} years")

Time taken for sample to reduce to half its original amount = 14 years


In [38]:
import numpy as np
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

np_co2 = np.array(co2_data)
co2_level = np_co2[:, 1]
for year in range(len(co2_level)):
    if (year == len(co2_level) - 1):  #required so that in the next step, the last year's [year+1] will not exceed the index limit of the list
        break
    if (co2_level[year + 1] - co2_level[year] >= 3):  #finding the two years with changes more than 3ppm
        print(f"Between {int(np_co2[year][0])} and {int(np_co2[year + 1][0])}")
        print(f"Change in CO2 concentration = {co2_level[year + 1] - co2_level[year]} ppm")

Between 2015 and 2016
Change in CO2 concentration = 3.410000000000025 ppm
